portfolio/src/main/webapp/script.js

```diff
@@ -60,6 +51,71 @@ async function getCommentFromFetch() {
         var para = document.createElement("p");
         var node = document.createTextNode(commentString);
         para.appendChild(node);
+        var button = document.createElement("button");
+        button.innerHTML = "Like";
+        button.type = "button";
+        const input = commentString.slice();
+        button.addEventListener("click", function () { likeComments(input) });
         commentContainer.appendChild(para);
+        commentContainer.appendChild(button);
     };
+}
+
+async function fetchCommentsWithStoredLimit() {
+    console.log("Try fetch");
+    const response = await fetch("/data");
+    const commentResponse = await response.json();
+    const comments = commentResponse.commentList;
+    var maxComments = parseInt(commentResponse.maxComments, 10);
+    if (Number.isNaN(maxComments)) {
+        console.log("You are trying to input a non-numeric data, please try again");
+        // Set the limit to default value
+        maxComments = 10;
+    }
+    const commentLimitContainer = document.getElementById("comment-limit");
+    commentLimitContainer.value = (maxComments.toString());
+    displayComments(maxComments, comments);
+}
+
+async function fetchCommentsWithInputLimit() {
+    console.log("Try fetch");
+    const response = await fetch("/data");
+    const commentResponse = await response.json();
+    const comments = commentResponse.commentList;
+    var maxComments = parseInt(document.getElementById("comment-limit").value, 10);
+    if (Number.isNaN(maxComments)) {
+        console.log("You are trying to input a non-numeric data, please try again");
+        // Set the limit to default value
+        maxComments = 10;
+    }
+    if (maxComments < 1) {
+        console.log("Please input a positive integer");
+        // Set the limit to default value
+        maxComments = 10;
+    }
+    displayComments(maxComments, comments);
+}
+
+async function deleteComments() {
+    console.log("Delete all comments in datastore");
+    const request = new Request("/comments", {method: "DELETE"});
+    const response = await fetch(request);
+    const commentContainer = document.getElementById("comment-container");
+    commentContainer.innerHTML = "";
+}
+
+async function likeComments(commentString) {
+    console.log("Like a certain comment");
+    const response = await fetch("/data");
+    const commentResponse = await response.json();
+    const comments = commentResponse.commentList;
+    var key;
+    for (var index = 0; index < comments.length; index++) {
+        if (commentString === comments[index].comment) {
+            key = comments[index].keyString
+        }
+    }
```

Comment:
Let me suggest another option here that will reduce round-trips to the server: take advantage having the key when you're creating the comment nodes in `displayComments`. Instead of having your click handlers for the Like button call this function with the comment string, call it with the key.

Reply:
That really helps! I didn't notice that after I added the keyString as a property of the entity, I can get the keyString of every comment in the `displayComments` function. I will improve this in the next commit.